# Bellabeat Smart Device Usage & Wellness Behaviour Analysis

## Stage 3: Process

### Processing Objective

The purpose of this stage is to inspect, clean and transform the Fitbit data so that it is accurate, consistent and ready for analysis.

The processing stage will include:

- Importing the required Python libraries
- Loading the selected Fitbit CSV files
- Inspecting the structure of the datasets
- Checking column names and data types
- Checking for missing values
- Checking for duplicate records
- Validating date fields
- Checking for unusual or invalid values
- Preparing cleaned datasets for analysis

In [1]:
import pandas as pd

In [2]:
import os

os.getcwd()

'c:\\Users\\OLUWATOSIN OLUWASEUN\\Downloads\\bellabeat-wellness-analysis\\notebooks'

In [3]:
daily_activity = pd.read_csv(
    "../data/raw/mturkfitbit_export_4.12.16-5.12.16/Fitabase Data 4.12.16-5.12.16/dailyActivity_merged.csv"
)

daily_activity.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [4]:
daily_activity.shape

(940, 15)

In [5]:
daily_activity.columns

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='str')

In [6]:
daily_activity.dtypes

Id                            int64
ActivityDate                    str
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object

In [7]:
daily_activity["ActivityDate"].head(10)

0    4/12/2016
1    4/13/2016
2    4/14/2016
3    4/15/2016
4    4/16/2016
5    4/17/2016
6    4/18/2016
7    4/19/2016
8    4/20/2016
9    4/21/2016
Name: ActivityDate, dtype: str

In [8]:
daily_activity["ActivityDate"].head(10)

0    4/12/2016
1    4/13/2016
2    4/14/2016
3    4/15/2016
4    4/16/2016
5    4/17/2016
6    4/18/2016
7    4/19/2016
8    4/20/2016
9    4/21/2016
Name: ActivityDate, dtype: str

In [9]:
daily_activity["ActivityDate"] = pd.to_datetime(
    daily_activity["ActivityDate"],
    format="%m/%d/%Y"
)

In [10]:
daily_activity["ActivityDate"].dtype

dtype('<M8[us]')

In [11]:
daily_activity.isnull().sum()

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

In [12]:
daily_activity.duplicated().sum()

np.int64(0)

In [13]:
daily_activity["Id"].nunique()

33

In [14]:
daily_activity.groupby("Id").size().sort_values()

Id
4057192912     4
2347167796    18
8253242879    19
3372868164    20
6775888955    26
7007744171    26
6117666160    28
6290855005    29
8792009665    29
5577150313    30
3977333714    30
1644430081    30
2320127002    31
1503960366    31
1624580081    31
1844505072    31
4388161847    31
4319703577    31
4020332650    31
2873212765    31
1927972279    31
2022484408    31
2026352035    31
5553957443    31
4445114986    31
4558609924    31
4702921684    31
6962181067    31
8053475328    31
7086361926    31
8378563200    31
8583815059    31
8877689391    31
dtype: int64

In [15]:
records_per_user = daily_activity.groupby("Id").size()

print("Minimum records for one user:", records_per_user.min())
print("Maximum records for one user:", records_per_user.max())

Minimum records for one user: 4
Maximum records for one user: 31


In [16]:
print("Start date:", daily_activity["ActivityDate"].min())
print("End date:", daily_activity["ActivityDate"].max())

Start date: 2016-04-12 00:00:00
End date: 2016-05-12 00:00:00


In [17]:
daily_activity[
    [
        "TotalSteps",
        "TotalDistance",
        "VeryActiveMinutes",
        "FairlyActiveMinutes",
        "LightlyActiveMinutes",
        "SedentaryMinutes",
        "Calories"
    ]
].min()

TotalSteps              0.0
TotalDistance           0.0
VeryActiveMinutes       0.0
FairlyActiveMinutes     0.0
LightlyActiveMinutes    0.0
SedentaryMinutes        0.0
Calories                0.0
dtype: float64

In [18]:
zero_counts = (daily_activity[
    [
        "TotalSteps",
        "TotalDistance",
        "VeryActiveMinutes",
        "FairlyActiveMinutes",
        "LightlyActiveMinutes",
        "SedentaryMinutes",
        "Calories"
    ]
] == 0).sum()

zero_counts

TotalSteps               77
TotalDistance            78
VeryActiveMinutes       409
FairlyActiveMinutes     384
LightlyActiveMinutes     84
SedentaryMinutes          1
Calories                  4
dtype: int64

In [19]:
daily_activity[daily_activity["Calories"] == 0]

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
30,1503960366,2016-05-12,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0
653,6290855005,2016-05-10,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0
817,8253242879,2016-04-30,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0
879,8583815059,2016-05-12,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,0,1440,0


In [20]:
daily_activity_clean = daily_activity[
    daily_activity["Calories"] > 0
].copy()

In [21]:
daily_activity_clean.shape

(936, 15)

### Daily Activity Data Cleaning Summary

The `dailyActivity_merged.csv` dataset initially contained 940 rows and 15 columns.

The following data-quality checks were completed:

- Confirmed 33 unique Fitbit user IDs.
- Converted `ActivityDate` from text to datetime format.
- Confirmed that there were no missing values.
- Confirmed that there were no exact duplicate rows.
- Verified that the dataset covers 12 April 2016 to 12 May 2016.
- Identified that users contributed between 4 and 31 daily records, indicating unequal data coverage across participants.
- Checked the main numeric activity variables for negative values and found none.
- Identified 4 records where `Calories` was equal to 0.
- These records also showed no meaningful recorded activity and were treated as incomplete tracker records.
- Removed the 4 zero-calorie records from the cleaned analytical dataset.

The cleaned daily activity dataset contains 936 rows and 15 columns.

The original raw CSV file remains unchanged.

In [22]:
sleep_day = pd.read_csv(
    "../data/raw/mturkfitbit_export_4.12.16-5.12.16/Fitabase Data 4.12.16-5.12.16/sleepDay_merged.csv"
)

sleep_day.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


In [23]:
sleep_day.shape

(413, 5)

In [24]:
sleep_day.columns.tolist()

['Id', 'SleepDay', 'TotalSleepRecords', 'TotalMinutesAsleep', 'TotalTimeInBed']

In [25]:
sleep_day.shape

(413, 5)

In [26]:
sleep_day.dtypes

Id                    int64
SleepDay                str
TotalSleepRecords     int64
TotalMinutesAsleep    int64
TotalTimeInBed        int64
dtype: object

In [27]:
sleep_day["SleepDay"].head(10)

0    4/12/2016 12:00:00 AM
1    4/13/2016 12:00:00 AM
2    4/15/2016 12:00:00 AM
3    4/16/2016 12:00:00 AM
4    4/17/2016 12:00:00 AM
5    4/19/2016 12:00:00 AM
6    4/20/2016 12:00:00 AM
7    4/21/2016 12:00:00 AM
8    4/23/2016 12:00:00 AM
9    4/24/2016 12:00:00 AM
Name: SleepDay, dtype: str

In [28]:
sleep_day["SleepDay"] = pd.to_datetime(
    sleep_day["SleepDay"],
    format="%m/%d/%Y %I:%M:%S %p"
)

In [29]:
sleep_day["SleepDay"].dtype

dtype('<M8[us]')

In [30]:
sleep_day.isnull().sum()

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

In [31]:
sleep_day.duplicated().sum()

np.int64(3)

In [32]:
sleep_day[sleep_day.duplicated(keep=False)].sort_values(
    by=["Id", "SleepDay"]
)

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
160,4388161847,2016-05-05,1,471,495
161,4388161847,2016-05-05,1,471,495
222,4702921684,2016-05-07,1,520,543
223,4702921684,2016-05-07,1,520,543
379,8378563200,2016-04-25,1,388,402
380,8378563200,2016-04-25,1,388,402


In [33]:
sleep_day_clean = sleep_day.drop_duplicates().copy()

In [34]:
sleep_day_clean.shape

(410, 5)

In [35]:
sleep_day_clean["Id"].nunique()

24

In [36]:
sleep_records_per_user = sleep_day_clean.groupby("Id").size()

print("Minimum sleep records for one user:", sleep_records_per_user.min())
print("Maximum sleep records for one user:", sleep_records_per_user.max())

Minimum sleep records for one user: 1
Maximum sleep records for one user: 31


In [37]:
print("Start date:", sleep_day_clean["SleepDay"].min())
print("End date:", sleep_day_clean["SleepDay"].max())

Start date: 2016-04-12 00:00:00
End date: 2016-05-12 00:00:00


In [38]:
sleep_day_clean[
    [
        "TotalSleepRecords",
        "TotalMinutesAsleep",
        "TotalTimeInBed"
    ]
].min()

TotalSleepRecords      1
TotalMinutesAsleep    58
TotalTimeInBed        61
dtype: int64

In [39]:
sleep_day_clean.sort_values(
    by="TotalMinutesAsleep"
).head(10)

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
338,7007744171,2016-05-01,1,58,61
123,4319703577,2016-04-21,1,59,65
65,2320127002,2016-04-23,1,61,69
164,4388161847,2016-05-09,1,62,65
282,5577150313,2016-05-05,1,74,78
365,8053475328,2016-05-07,1,74,75
110,4020332650,2016-04-16,1,77,77
337,7007744171,2016-04-16,1,79,82
150,4388161847,2016-04-22,1,82,85
172,4445114986,2016-04-17,1,98,107


In [40]:
(sleep_day_clean["TotalMinutesAsleep"] > sleep_day_clean["TotalTimeInBed"]).sum()

np.int64(0)

In [41]:
sleep_day_clean.duplicated(
    subset=["Id", "SleepDay"]
).sum()

np.int64(0)

### Sleep Data Cleaning Summary

The `sleepDay_merged.csv` dataset initially contained 413 rows and 5 columns.

The following data-quality checks were completed:

- Converted `SleepDay` from text to datetime format.
- Confirmed there were no missing values.
- Identified 3 exact duplicate records.
- Removed the 3 duplicate records.
- Confirmed 24 unique Fitbit users have sleep data.
- Identified that users contributed between 1 and 31 sleep records.
- Verified that the sleep data covers 12 April 2016 to 12 May 2016.
- Inspected unusually short sleep records and retained them because they remained logically consistent.
- Confirmed that `TotalMinutesAsleep` never exceeded `TotalTimeInBed`.
- Confirmed that each remaining user-date combination is unique.

The cleaned sleep dataset contains 410 rows and 5 columns.

The original raw CSV file remains unchanged.

In [42]:
daily_activity_clean.to_csv(
    "../data/processed/daily_activity_clean.csv",
    index=False
)

sleep_day_clean.to_csv(
    "../data/processed/sleep_day_clean.csv",
    index=False
)

In [43]:
check_daily = pd.read_csv("../data/processed/daily_activity_clean.csv")
check_sleep = pd.read_csv("../data/processed/sleep_day_clean.csv")

print("Daily activity shape:", check_daily.shape)
print("Sleep data shape:", check_sleep.shape)

Daily activity shape: (936, 15)
Sleep data shape: (410, 5)
